### DISCUSSION TITLE
Using useAtomCallback in micro-frontend architecture

### DISCUSSION THREAD

**Discussion by soonki-98** [pmndrs/jotai#2130] [Q&A]
when I use `useAtomCallback` for change specific atom named `modalListAtom`, Component that subscribe `modalListAtom` 
does not re-rendered.

I wonder if the micro front-end through `module federation` affects the using useAtomCallback of Jotai.
Am I using it wrong or is this the right move?

below is my example. (actual code is much complicated)

---
### packages/lib/useModal.ts
```tsx
import modalListAtom from "./modalListAtom.ts"
import { useAtomCallback } from "jotai/utils";
import { useCallback } from "react";

const useModal = () => {
     const showModal = useAtomCallback(useCallback((_, set, params) => {
                set(modalListAtom, (prev) => [...prev, params]);
     });

    return { showModal };
}
```
---
### packages/lib/ModalContainer.tsx
```tsx 
import modalListAtom from "./modalListAtom.ts"
import { useAtomValue } from "jotai";

const ModalContainer = () => {
     const modalList = useAtomValue(modalListAtom);
     console.log(modalList); // this log always print initial state
     return React.createPortal({ modalList.map(( modal ) => <Modal {...modal} /> }, document.body);
}
```
---
### packages/apps/app1/App.tsx
```tsx 
import { Provider, createStore } from "jotai";
import { ModalContainer } from "lib/Modal";

const App2 = React.lazy(() => import("apps/app2/App.tsx"));
const store = createStore();

const App = () => {
    return (
     <Provider store={store}>
         <React.Suspense>
            <ModalContainer />
            <Child />
           <App2 />
         </React.Suspense>
    </Provider>
   )
}
```
---
### packages/apps/app1/Child.tsx
```tsx
// packages/apps/app1/Child.tsx
import { useModal } from "lib/Modal";

const Child = () => {
     const { showModal } = useModal();
     return <button onClick={showModal}>open modal</button>
}
```

**Comment by dai-shi** [comment]
> I wonder if the micro front-end through `module federation` affects the using useAtomCallback of Jotai.

I'm not exactly sure. Would you try a small example with https://csb.jotai.org (no module federation) and see if it works?

**Reply by soonki-98** [reply]
I already tried. It works well with https://csb.jotai.org/ but it doesn't works with my app (module federation).

**Reply by dai-shi** [reply]
Does it only fail with `useAtomCallback`?
Does it work with `useSetAtom`?

**Reply by soonki-98** [reply]
yes `useSetAtom` does works

**Reply by soonki-98** [reply]
Is this problem related to the return value of 'useAtomCallback' being Promise?

**Reply by dai-shi** [reply]
> yes `useSetAtom` does works

Interesting.

> Is this problem related to the return value of 'useAtomCallback' being Promise?

Wait, which version of Jotai are you using?

**Reply by soonki-98** [reply]
my Jotai version is `2.1.0`

**Reply by dai-shi** [reply]
Hm, useAtomCallback implementation is pretty simple.

https://github.com/pmndrs/jotai/blob/2526039ea4da082749adc8a449c33422c53d9819/src/react/utils/useAtomCallback.ts#L8-L17

I wonder if you dig into it and investigate what cause wrong. Find a difference between with module federation and without it.

**Reply by soonki-98** [reply]
Okay thank you for your time. I'll looking for the problems with my app

**Comment by BlessYAHU** [comment]
Add your usemodal.ts to shared decencies in the modulefederation section of webpack.config, WITH the singleton set to true:
"Some libraries use a global internal state (e.g. react, react-dom). Thus, it is critical to have only one instance of the library running at a time."  Via https://webpack.js.org/plugins/module-federation-plugin/#sharing-hints

This allowed two remotes I had to get the atom changes. Add this config to the remote and host webpack configs.

**Reply by soonki-98** [reply]
I don't understand about "usemodal.take" can you explain more?
In my project, I already added 'jodai' to shared dependencies 
```typescript
module.exports = {
  // ...other webpack configuration
  plugins: [
    new ModuleFederationPlugin({
      // ...other federation configuration
      shared: {
        jotai: {
          singleton: true,
          requiredVersion: package.dependencies['jotai']
        },
      },
    }),
  ],
};
```

**Reply by BlessYAHU** [reply]
> I don't understand about "usemodal.take" can you explain more?
> 
> In my project, I already added 'jodai' to shared dependencies 
> 

Usermodal.take was a typo. You also need to add your usermodal.ts module to the shared dependencies for the remotes that need access to it.

**Comment by soonki-98** [comment]
I solve this problem with using store options.
I was resolve by referring to the [discussion](https://github.com/pmndrs/jotai/discussions/2044).
This problem was occured because of multi store issue. so I created single store to solve and it works.
